CRM/Event/Form/Participant.php

```diff
@@ -779,16 +779,21 @@ public static function formRule($values, $files, $self) {
       $event->id = $eventId;
       $event->find(TRUE);
 
-      if (!$event->allow_same_participant_emails && !empty($contactId) && !empty($eventId)) {
-        $cancelledStatusID = CRM_Core_PseudoConstant::getKey('CRM_Event_BAO_Participant', 'status_id', 'Cancelled');
-        $dupeCheck = new CRM_Event_BAO_Participant();
-        $dupeCheck->contact_id = $contactId;
-        $dupeCheck->event_id = $eventId;
-        $dupeCheck->whereAdd("status_id != {$cancelledStatusID} ");
-        $dupeCheck->find(TRUE);
-        if (!empty($dupeCheck->id)) {
-          $errorMsg['event_id'] = ts('This contact has already been assigned to this event.');
-        }
+      if (
+        !$event->allow_same_participant_emails
+        && !empty($contactId)
+        && !empty($eventId)
+        && CRM_Event_BAO_Participant::exists(
+          $contactId,
+          $eventId,
+          FALSE,
+          TRUE,
+          ['Cancelled'],
```

Comment:
Do we explicitly have to exclude `Cancelled` instead of relying on the status not having `is_counted` and passing `TRUE` for `$onlyCounted`?

Reply:
So I think the assumption here is that a single contact should only ever have a single non cancelled participant registration for an event. Even if that registration is on the waitlist, pending from waitlist or others so I think the point here that we want to include any registrations that are pending / on waitlist as well as the positive statuses.

Reply:
Well, that's what the old code was doing (and including `is_test` participants btw.), so I tried to retain that with the new code. The question is if that's a sane assumption, considering participant status are configurable with `is_counted` etc. - think of a negative status similar to `Cancelled` or `Cancelled` being set to `is_counted` … but maybe too broad of a question to discuss for simple refactoring.

Reply:
As per discussion with @colemanw and @JoeMurray, we should rely on `is_counted` and get rid of the status name filter (i.e. not passing `['Cancelled']` here and pass `$onlyCounted` as `TRUE`).